Nextcloud

Deric-W opened the pull request:
Since nextcloud has a [hardening guide](https://docs.nextcloud.com/server/27/admin_manual/installation/harden_server.html#setup-a-filter-and-a-jail-for-nextcloud) which contains a fail2ban filter I decided that it would be useful for including it here (split into two filters to allow different jails).

Since the original filter fails when using the systemd journal I modified it to accept the prefix line (optional) from `common.conf`.

Before submitting your PR, please review the following checklist:

- [x] **CHOOSE CORRECT BRANCH**: if filing a bugfix/enhancement
      against certain release version, choose `0.9`, `0.10` or `0.11` branch,
      for dev-edition use `master` branch
- [x] **CONSIDER adding a unit test** if your PR resolves an issue
- [x] **LIST ISSUES** this PR resolves
- [x] **MAKE SURE** this PR doesn't break existing tests
- [x] **KEEP PR small** so it could be easily reviewed.
- [x] **AVOID** making unnecessary stylistic changes in unrelated code
- [x] **ACCOMPANY** each new `failregex` for filter `X` with sample log lines
      within `fail2ban/tests/files/logs/X` file

Deric-W commented:
Do substitutions have to be enclosed in `()` if you want to apply `?` to them?

sebres commented on fail2ban/tests/files/logs/nextcloud-auth:
```diff
@@ -0,0 +1,8 @@
+# failJSON: { "time": "2023-09-24T22:34:37.0+02:00", "match": true , "host": "141.30.226.119" }
```
Our filter test factory already works in CE(S)T time zone, so normally it would be enough to give the TZ-offset directly in the time.
So for `"time":"2023-09-24T20:34:37+00:00"` it would be `2023-09-24T22:34:37`...
But OK, we can accept this change.

Deric-W replied:
I did not know that the test factory works in CET time and assumed it just used the local time of my computer.
In this case I can revert the change to the test factory.

Neustradamus commented:
@Deric-W: Have you seen comments on your PR?

Deric-W commented:
Sorry for the very long delay, I totally forgot about this PR.
I will address them when I am back home.

Deric-W commented on config/filter.d/nextcloud-auth.conf:
```diff
@@ -0,0 +1,18 @@
+# Fail2Ban filter file for Nextcloud login failures
+#
+# Author: Eric Wolf
```
I assumed that the author is the person who created the filter file.
Since the regex is from the hardening guide (which after some investigation turns out to be based on #2485) I would put your name there with the name of the person who added the trusted domain error extension in nextcloud/documentation#2421.

Deric-W commented on config/filter.d/nextcloud-auth.conf:
```diff
@@ -0,0 +1,18 @@
+# Author: Eric Wolf
+#
+
+[INCLUDES]
+
+# Read common prefixes
+before = common.conf
+
+after = nextcloud-auth.local
```
Including `common.conf` is required since it provides `__prefix_line`.
I added the `%(__prefix_line)s?` part to allow using it when using the systemd journal for logging since it would not match without it.
The [example](https://fail2ban.readthedocs.io/en/latest/filters.html#filter-file) in the filter guide made me assume that the `after` include was required, turns out I was wrong.

Deric-W commented on config/filter.d/nextcloud-auth.conf:
```diff
@@ -0,0 +1,18 @@
+# based on https://docs.nextcloud.com/server/27/admin_manual/installation/harden_server.html#setup-a-filter-and-a-jail-for-nextcloud
+_groupsre = (?:(?:,?\s*"\w+":(?:"[^"]+"|\w+))*)
+failregex = ^%(__prefix_line)s?\{%(_groupsre)s,?\s*"remoteAddr":"<HOST>"%(_groupsre)s,?\s*"message":"Login failed:
+datepattern = ,?\s*"time"\s*:\s*"%%Y-%%m-%%d[T ]%%H:%%M:%%S(%%z)?"
```
I think the different formats are from a predecessor of Nextcloud called ownCloud and the versions before 18 which apparently had a different timestamp.
The pull request which added it to the hardening guide (see my comment on the author) mentions it working on Nextcloud 18.
Since it still works on Nextcloud 28 and the filter is used by [NextcloudPi](https://github.com/nextcloud/nextcloudpi/blob/f64af68384a359a7c008deb2ef2ab8940f15ba84/bin/ncp/SECURITY/fail2ban.sh#L63C1-L71C14) it seems that it is still valid.

I anchored the time pattern by prefixing it with `^%(__prefix_line)s?\{%(_groupsre)s`.

Deric-W commented:
I noticed that the group regex `(?:(?:,?\s*"\w+":(?:"[^"]+"|\w+))*)` is vulnerable against input containing quoted quotation marks (\\") which cause it to think that a group has ended and fail the match when a group like `...,,"url":"/login\"",...` would appear before the part we want to match.
I am not sure if this is possible in practice since all groups we want to match in the example log lines are either not directly attacker controlled or url encoded.

I fixed it by modifying the group regex with `(?:(?:,?\s*"\w+":(?:"(?:[^"\\]|\\"|\\\\)*"|\w+))*)`, can you spot any problems with this regex?

sebres commented:
> can you spot any problems with this regex?

yes, because now `[^"\\]` would stop on backslash too, and it can be used to escape any char, so theoretically it must be:
```diff
- (?:(?:,?\s*"\w+":(?:"(?:[^"\\]|\\"|\\\\)*"|\w+))*)
+ (?:(?:,?\s*"\w+":(?:"(?:[^"\\]+|\\.)*"|\w+))*)
```
(also note the `+` quantifier after `[^"\\]` - that makes regular case much faster)

Deric-W commented:
Thanks for catching the mistake 👍
The performance optimization however seems to cause exponential backtracking which makes the test suite hang.
I removed the `+` quantifier and moved the date pattern and some parts to a `nextcloud-common.conf` (based on `apache-common.conf`).

yitzhaq commented:
One problem with using this filter - though not a problem with the regex or anything here per se - is that it is prone to false positives because the logs don't actually say anything about *why* the login failed. It could be because of a genuine invalid login and/or brute force attempt, sure, but the logs will look identical if there is some sort of problem with the backend authentication provider, triggering a slew of false alarms. The logged statement is true in both cases, in that the login *did* indeed fail, but some more detail (like what eg. Dovecot does) would allow to differentiate between these two very different problems, and make it possible to adjust the filter regex accordingly to achieve higher accuracy.

Not something that can be fixed here of course, as this is clearly a Nextcloud bug (or at least I would consider it to be one), but perhaps something to note as a comment, a possible gotcha worth being aware of? I at least had to disable my Nextcloud jail because I could not find any way to prevent this from happening.

anantakrishna commented:
I guess this PR makes #1913 outdated. Does it?

Deric-W commented:
I think so, but I noticed that the mentioned PR seems to expect a different nextcloud log format (probably because it is 7 years old).

joshtrichards commented:
@yitzhaq Feel free to bring up your suggestions at the [Nextcloud Server](https://github.com/nextcloud/server) (if logging behavior changes are required) and/or [Documentation](https://github.com/nextcloud/documentation) (if rule changes are required) repositories. :-)
